### PR TITLE
WWST-2225 [China] Ezex Temp/Humidity Sensor

### PR DESCRIPTION
**Pull request by FenMei007S**
@tpmanley  DTH for ezex sensor. this device is for kroea market

### CODE REVIEW

File: devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+	// Device-Watch allows 2 check-in misses from device + ping (plus 1 min lag time)
+	// enrolls with default periodic reporting until newer 5 min interval is confirmed
+	sendEvent(name: "checkInterval", value: 2 * 60 * 60 + 1 * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID])
+
+	log.debug "Configuring Reporting and Bindings."
+
+	// temperature minReportTime 30 seconds, maxReportTime 5 min. Reporting interval if no activity
+	return refresh() +
+		  zigbee.configureReporting(0x0405, 0x0000, DataType.UINT16, 30, 3600, 100, [destEndpoint: 0x02]) +
+                  zigbee.configureReporting(0x0402, 0x0000, DataType.UINT16, 30, 3600, 100, [destEndpoint: 0x02])
```
Comment (varzac) [inline devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy]: should this be endpoint 1?  in the readAttr in refresh temperature is endpoint 1, but here it is endpoint 2

File: devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+def configure() {
+	// Device-Watch allows 2 check-in misses from device + ping (plus 1 min lag time)
+	// enrolls with default periodic reporting until newer 5 min interval is confirmed
+	sendEvent(name: "checkInterval", value: 2 * 60 * 60 + 1 * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID])
+
+	log.debug "Configuring Reporting and Bindings."
+
+	// temperature minReportTime 30 seconds, maxReportTime 5 min. Reporting interval if no activity
```
Comment (varzac) [inline devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy]: This comment is not accurate.

File: devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+								[value: 84, color: "#f1d801"],
+								[value: 95, color: "#d04e00"],
+								[value: 96, color: "#bc2323"]
+						]
+			}
+		}
+		valueTile("humidity", "device.humidity", inactiveLabel: false, width: 2, height: 2) {
+			state "humidity", label: '${currentValue}% humidity', unit: ""
+		}
+		valueTile("battery", "device.battery", decoration: "flat", inactiveLabel: false, width: 2, height: 2) {
```
Comment (greens) [inline devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy]: Device doesn't support battery capability. Remove this tile.

File: devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy [hunk 4]
```diff
@@ -0,0 +1,129 @@
+/**
+ *  SmartSense Temp/Humidity Sensor
+ *
+ *  Copyright 2014 SmartThings
```
Comment (greens) [inline devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy]: update year

File: devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
+ *  for the specific language governing permissions and limitations under the License.
+ *
+ */
+import physicalgraph.zigbee.zcl.DataType
+
+metadata {
+	definition(name: "eZEX Temperature & Humidity Sensor", namespace: "smartthings", author: "SmartThings", runLocally: false, minHubCoreVersion: '000.017.0012', executeCommandsLocally: false, mnmn:"SmartThings", vid:"generic-humidity") {
```
Comment (greens) [inline devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy]: I think the words in the DTH name needs to match the filename

this should be `eZEX Temp Humidity Sensor` or `eZEX Temp/Humidity Sensor`. `eZEX Temp & Humidity Sensor` might work.

File: devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+		capability "Configuration"
+		capability "Temperature Measurement"
+		capability "Relative Humidity Measurement"
+		capability "Sensor"
+		
+                fingerprint profileId: "0104", inClusters: "0000,0003,0402,0405,0500", outClusters: "0019", model: "E282-KR0B0Z1-HA", deviceJoinName: "eZEX Temperature & Humidity Sensor(AC Type)"
+	}
+
+	simulator {
```
Comment (greens) [inline devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy]: Remove this section. 

`FC45` is a custom SmartThings cluster. This device does not support these messages.

File: devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+/**
+ * PING is used by Device-Watch in attempt to reach the Device
+ * */
+def ping() {
+	return refresh()
+}
+
+def refresh() {
+	log.debug "refresh temperature, humidity"
+	return zigbee.readAttribute(0x0402, 0x0000)+
```
Comment (greens) [inline devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy]: replace magic numbers with

`zigbee.RELATIVE_HUMIDITY_CLUSTER` and `zigbee.TEMPERATURE_MEASUREMENT_CLUSTER`

Comment (greens) [reply devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy]: you forgot these

File: devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed
+ *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
+ *  for the specific language governing permissions and limitations under the License.
+ *
+ */
+import physicalgraph.zigbee.zcl.DataType
+
+metadata {
+	definition(name: "eZEX Temp & Humidity Sensor", namespace: "smartthings", author: "SmartThings", runLocally: false, minHubCoreVersion: '000.017.0012', executeCommandsLocally: false, mnmn:"SmartThings", vid:"generic-humidity") {
```
Comment (tpmanley) [inline devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy]: Can you remove `runLocally: false, minHubCoreVersion: '000.017.0012', executeCommandsLocally: false,`? We usually don't put that stuff in if it doesn't run locally

File: devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+		capability "Configuration"
+		capability "Temperature Measurement"
+		capability "Relative Humidity Measurement"
+		capability "Sensor"
+		
+                fingerprint profileId: "0104", inClusters: "0000,0003,0402,0405,0500", outClusters: "0019", model: "E282-KR0B0Z1-HA", deviceJoinName: "eZEX Temperature & Humidity Sensor(AC Type)"
```
Comment (tpmanley) [inline devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy]: Check the indentation on this line and the rest of the file to make sure its consisten

Comment (tpmanley) [reply devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy]: Also I suggest a space after `Sensor`

File: devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+ * */
+def ping() {
+	return refresh()
+}
+
+def refresh() {
+	log.debug "refresh temperature, humidity"
+	return zigbee.readAttribute(zigbee.RELATIVE_HUMIDITY_CLUSTER, 0x0000)+
+		  zigbee.readAttribute(zigbee.TEMPERATURE_MEASUREMENT_CLUSTER, 0x0000)
+
```
Comment (tpmanley) [inline devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy]: nit: extra newline

File: devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+def configure() {
+	// Device-Watch allows 2 check-in misses from device + ping (plus 1 min lag time)
+	sendEvent(name: "checkInterval", value: 2 * 60 * 60 + 1 * 60, displayed: false, data: [protocol: "zigbee", hubHardwareId: device.hub.hardwareID])
+
+	log.debug "Configuring Reporting and Bindings."
+
+	// temperature minReportTime 30 seconds, maxReportTime 1 hour. Reporting interval if no activity
+	return refresh() +
+		  zigbee.configureReporting(zigbee.RELATIVE_HUMIDITY_CLUSTER, 0x0000, DataType.UINT16, 30, 3600, 100) +
+                  zigbee.configureReporting(zigbee.TEMPERATURE_MEASUREMENT_CLUSTER, 0x0000, DataType.UINT16, 30, 3600, 100)
```
Comment (tpmanley) [inline devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy]: The indentation looks off here too

File: devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+/**
+ * PING is used by Device-Watch in attempt to reach the Device
+ * */
+def ping() {
+	return refresh()
+}
+
+def refresh() {
+	log.debug "refresh temperature, humidity"
+	return zigbee.readAttribute(zigbee.RELATIVE_HUMIDITY_CLUSTER, 0x0000)+
```
Comment (tpmanley) [inline devicetypes/smartthings/ezex-temp-humidity-sensor.src/ezex-temp-humidity-sensor.groovy]: nit: space before `+`

**Review by tpmanley — APPROVED** [review]

**Pull request merged**